servers/jm_utils/jm_utils/page_tokens.py

```diff
@@ -1,44 +1,137 @@
 
     Args:
-      offset: (int) index into the overall list of jobs matching the query
+        dictionary: (dict) Dictionary to basee64-encode
+    
     Returns:
-      (string) encoded page token representing a page of jobs
+        (string) encoded page token representing a page of jobs
     """
-    s = json.dumps({
-        'of': offset,
-    })
     # Strip ugly base64 padding.
-    return base64.urlsafe_b64encode(s).rstrip('=')
+    return base64.urlsafe_b64encode(json.dumps(dictionary)).rstrip('=')
 
 
-def decode(token):
-    """Decode the jobs pagination token.
+def _decode(token):
+    """Decodes the jobs pagination token.
 
     Args:
-      token: (string) base64 encoded JSON pagination token
+        token: (string) Base64 encoded JSON pagination token
 
     Returns:
-      (number) the pagination offset, defaults to 0 if token is None
+        (dict) The token dictionary representing a page of jobs.
     """
     if not token:
-        return 0
+        return None
     # Pad the token out to be divisible by 4.
     padded_token = token + '=' * (4 - (len(token) % 4))
     decoded_token = base64.urlsafe_b64decode(padded_token)
     token_dict = json.loads(decoded_token)
+    if not token_dict or not isinstance(token_dict, dict):
+        raise ValueError('Invalid pagination token: {}').format(token_dict)
+    return token_dict
+
+
+def encode_offset(offset):
+    """Encodes an offset integer into a jobs pagination token.
+    
+    Args:
+      offset: (int) index into the overall list of jobs matching the query
+    
+    Returns:
+      (string) encoded page token representing a page of jobs
+    """
+    if offset == None:
+        return None
+
+    if not isinstance(offset, numbers.Number) or offset <= 0:
+        raise ValueError('Invalid offset must integer > 0: {}'.format(offset))
+    return _encode({'of': offset})
+
+
+def encode_created_before(created_before, offset_id=None):
+    """Encodes a created_before and optional offset job ID into a jobs pagination token.
+    
+    Args:
+      created_before: (datetime) The create-time of the first job to include in the associated page.
+      offset_id: (str) The unique job-id of the first job to include in the associated page. Needed
+        only if the previous page ends with a job containing the same create-time.
+    
+    Returns:
+      (string) encoded page token representing a page of jobs.
+    """
+    if not created_before:
+        return None
+
+    if not isinstance(created_before, datetime.datetime):
+        raise ValueError(
+            'Invalid create time must be datetime: {}'.format(created_before))
+    if offset_id and not isinstance(offset_id, basestring):
+        raise ValueError(
+            'Invalid offset id must be string: {}'.format(offset_id))
+
+    epoch = datetime.datetime.utcfromtimestamp(0)
+    seconds_epoch = int((created_before - epoch).total_seconds())
+    token_dict = {'created-before': seconds_epoch}
```

Comment:
Please use shorter key names (1-2 chars ideally), each character here adds ~1.33 characters to the token length

Reply:
Right makes sense why offset is `of` now 😄